Implement layout support for position-area containing blocks

fantasai opened the pull request:
#### f1c4072ee9adb4482a8f232387b6e7f2c31a2a73
<pre>
Implement layout support for position-area containing blocks
<a href="https://bugs.webkit.org/show_bug.cgi?id=275447">https://bugs.webkit.org/show_bug.cgi?id=275447</a>
<a href="https://rdar.apple.com/129789558">rdar://129789558</a>

Reviewed by Alan Baradlay.

Adds consideration of position-area into our positioned layout containing block
calculations in the PositionedLayoutConstraints constructor.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-014-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-basic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-change-fallbacks-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-change-try-rule-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-intermediate-ignored-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-align-justify-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-align-justify-wm-dir-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-anchor-outside-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-anchor-partially-outside-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-basic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-in-grid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-in-position-try-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-value-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-with-insets-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-wm-dir-expected.txt:
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::PositionedLayoutConstraints::needsAnchor const): Add check for position-area.
(WebCore::RenderBox::PositionedLayoutConstraints::PositionedLayoutConstraints): Add call to computeAnchorGeometry().
(WebCore::RenderBox::PositionedLayoutConstraints::computeAnchorGeometry): Calculate the anchor area and derive the resulting containing block geometry.
(WebCore::RenderBox::PositionedLayoutConstraints::adjustForPositionArea): Helper function to apply position-area to a containing block + anchor area.

Canonical link: <a href="https://commits.webkit.org/291154@main">https://commits.webkit.org/291154@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7998ebcfac309c51ac03e0998c4add02e7697f07

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/92014 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/11545 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/1095 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/96937 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/42604 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/94064 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/11880 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/20037 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/70597 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/28116 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-anchor-position/position-visibility-anchors-visible-non-intervening-container.html (failure)") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/95015 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/9046 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/83341 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50925 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/8778 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/945 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/41819 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/79097 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/939 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/98973 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/19124 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/14148 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/79624 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/19376 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/79196 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78856 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19550 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/23388 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/708 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/12159 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/19105 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/24311 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/18802 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/22260 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/20550 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -3944,15 +3943,85 @@ struct RenderBox::PositionedLayoutConstraints {
     void convertLogicalTopValue(LayoutUnit& logicalTopValue, const RenderBox* child, const LayoutUnit logicalHeightValue, const bool logicalTopIsAuto, const bool logicalBottomIsAuto) const;
 };
 
+void RenderBox::PositionedLayoutConstraints::computeAnchorGeometry(const RenderBox& self)
```
I don't think **_self_** is what we normally use (or at all)

fantasai replied:
I originally used `box`, but eventually decided to align with the `self-` keywords in the alignment and `position-area` properties. I'm fine to use whichever, lmk what to do.

(I think `child` is confusing though since we're operating inside the box's own methods, not operating on its child.)

alanbaradlay replied:
@smfr wdyt?

fantasai replied:
(Note to Simon: EWS is overloaded this week, so gonna flag this for merging as-is and follow up later.)

alanbaradlay commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -3944,15 +3943,85 @@ struct RenderBox::PositionedLayoutConstraints {
+    marginPercentageBasis = adjustForPositionArea(inlineContainingBlock, inlineAnchorArea, inlineAxis, self.style()).size();
+}
+
+LayoutRange RenderBox::PositionedLayoutConstraints::adjustForPositionArea(const LayoutRange rangeToAdjust, const LayoutRange anchorArea, const BoxAxis containerAxis, const RenderStyle& style)
+{
```
I am not totally sure how this works but adopting this new coding style (`const` on pass by value) might be a good topic to discuss on the webkit mailing list first.

fantasai replied:
Should I remove them before landing then?

alanbaradlay replied:
If this PR touched tons of functions and files I would say yes, but since this just a small, local change, I think it's fine even if you may need to revert it in the future. Go with what works best for you.

fantasai committed:
Implement layout support for position-area containing blocks
https://bugs.webkit.org/show_bug.cgi?id=275447
rdar://129789558

Reviewed by Alan Baradlay.

Adds consideration of position-area into our positioned layout containing block
calculations in the PositionedLayoutConstraints constructor.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-014-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-basic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-change-fallbacks-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-change-try-rule-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-intermediate-ignored-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-align-justify-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-align-justify-wm-dir-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-anchor-outside-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-anchor-partially-outside-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-basic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-in-grid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-in-position-try-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-value-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-with-insets-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-wm-dir-expected.txt:
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::PositionedLayoutConstraints::needsAnchor const): Add check for position-area.
(WebCore::RenderBox::PositionedLayoutConstraints::PositionedLayoutConstraints): Add call to computeAnchorGeometry().
(WebCore::RenderBox::PositionedLayoutConstraints::computeAnchorGeometry): Calculate the anchor area and derive the resulting containing block geometry.
(WebCore::RenderBox::PositionedLayoutConstraints::adjustForPositionArea): Helper function to apply position-area to a containing block + anchor area.

Canonical link: https://commits.webkit.org/291154@main
